src/lib/pubkey/hss_lms/lms.cpp

```diff
@@ -0,0 +1,432 @@
+            case 15:
+               return LMS_Algorithm_Type::SHAKE_M24_H15;
+            case 20:
+               return LMS_Algorithm_Type::SHAKE_M24_H20;
+            case 25:
+               return LMS_Algorithm_Type::SHAKE_M24_H25;
+            default:
+               throw Decoding_Error("Unsupported height for hash function");
+         }
+      }
+      throw Decoding_Error("Unsupported hash function");
+   }(hash_name, height);
+
+   return LMS_Params(type, hash_name, height);
+}
+
+LMS_Params::LMS_Params(LMS_Algorithm_Type algorithm_type, std::string_view hash_name, uint8_t h) :
+      m_algorithm_type(algorithm_type), m_h(h), m_hash_name(hash_name) {
+   const auto hash = HashFunction::create_or_throw(m_hash_name);
+   m_m = hash->output_length();
+}
+
+LMS_PublicKey LMS_PrivateKey::sign_and_get_pk(StrongSpan<LMS_Signature_Bytes> out_sig,
+                                              LMS_Tree_Node_Idx q,
+                                              const LMS_Message& msg) const {
+   // Pre-alloc space for the signature
+   BOTAN_ARG_CHECK(out_sig.size() == LMS_Signature::size(lms_params(), lmots_params()), "Invalid output buffer size");
+
+   BufferStuffer sig_stuffer(out_sig);
+   sig_stuffer.append(store_be(q));
+   const LMOTS_Private_Key lmots_sk(lmots_params(), identifier(), q, seed());
+   lmots_sk.sign(sig_stuffer.next<LMOTS_Signature_Bytes>(LMOTS_Signature::size(lmots_params())), msg);
+   sig_stuffer.append(store_be(lms_params().algorithm_type()));
+   const auto auth_path_buffer = sig_stuffer.next<LMS_AuthenticationPath>(lms_params().m() * lms_params().h());
+
+   BOTAN_ASSERT_NOMSG(sig_stuffer.full());
+
+   TreeAddress lms_tree_address(lms_params().h());
+   LMS_Tree_Node pk_buffer(lms_params().m());
+   lms_treehash(StrongSpan<LMS_Tree_Node>(pk_buffer.get()), auth_path_buffer, q, *this);
+
+   return LMS_PublicKey(lms_params(), lmots_params(), identifier(), std::move(pk_buffer));
+}
+
+LMS_PublicKey LMS_PublicKey::from_bytes_or_throw(BufferSlicer& slicer) {
+   size_t total_remaining_bytes = slicer.remaining();
+   // Alg. 6. 1. (4 bytes are sufficient until the next check)
+   if(total_remaining_bytes < sizeof(LMS_Algorithm_Type)) {
+      throw Decoding_Error("Too few bytes while parsing LMS public key.");
+   }
+   // Alg. 6. 2.a.
+   auto lms_type = load_be<LMS_Algorithm_Type>(slicer.take<sizeof(LMS_Algorithm_Type)>());
+   // Alg. 6. 2.c.
+   auto lms_params = LMS_Params::create_or_throw(lms_type);
+   // Alg. 6. 2.d.
+   if(total_remaining_bytes < size(lms_params)) {
+      throw Decoding_Error("Too few bytes while parsing LMS public key.");
+   }
+   // Alg. 6. 2.b.
+   auto lmots_type = load_be<LMOTS_Algorithm_Type>(slicer.take<sizeof(LMOTS_Algorithm_Type)>());
+   auto lmots_params = LMOTS_Params::create_or_throw(lmots_type);
+
+   if(lms_params.hash_name() != lmots_params.hash_name()) {
+      throw Decoding_Error("No support for HSS-LMS instances with multiple hash functions.");
```

Comment:
I wanted to stay consistent so that every parsing error of the public key creates a `Decoding_Error`. In my interpretation, `Not_Implemented` is used when a feature is missing but could be added in the future. However, at this point, we decided strictly against the support of multiple hash functions per tree (which is also not recommended in the respective documents). 
But I'm open for discussion.

Reply:
Yeah, sounds good to me. It's a proper nit-pick anyway.